plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -126,14 +126,7 @@ prerequisite. The CEK machine correctly handles name shadowing.
 -}
 
 data ExBudgetCategory fun
-    = BConst
```

Comment:
I don't know how that could work with doing the slippage: the whole point is that we only have an integer counter for the number of steps. Keeping e.g. a list of the types so we could account for them would likely destroy that advantage. I don't know what to do about that.

Reply:
Indeed, the assumption that all the steps have the same cost is fairly fundamental to this whole approach... I'm not sure how to reconcile that.

Reply:
>  Keeping e.g. a list of the types so we could account for them would likely destroy that advantage. I don't know what to do about that.

Could we do something like making `maybeSpendBudget` a parameter to `enterComputeCek` which takes an argument of type `ExBudgetCategory` (or even just an integer tag): in counting mode that function would ignore the argument and just do what it's doing now, but in tallying mode it would update a map as before?  That's kind of unpleasant and maybe it would lose all the efficiency that the current version gives us.  The problem is that in tallying mode we do need to do something different for each node type, but that requires us to have some extra information that we don't need in counting mode, so it's difficult to do that uniformly in one program.  

Maybe we could write a separate stripped-down evaluator that does nothing but tallying, but that would bring problems of its own.

Reply:
Could we have a `PrimArray` with one element for each constructor, increment the appropriate slot at each step, and then spend a weighted sum of those every 100th step and reset the array to zero?  To count the number of times each node type is seen you'd never reset the array, or have another array that contains the running totals.